---
title: "Whole Genome Analysis: Measuring Methylation from Sequencing"
author: "Quynh Nhu Nguyen"
date: "March 4th 2023"
output: html_document
---
### Question 1:

Reduced Representation Bisulfite Sequencing or RRBS is an experimental technique widely used to manipulate the regions of the genome we measure. An enzyme is used to cut DNA at CCGG and the general idea is to filter out small or large molecules once DNA is cut. We can use Bioconductor tools to predict the size of these regions.

Load the genome package and create an object with the sequence for chr22:

```{r}
library("BSgenome.Hsapiens.UCSC.hg19")
chr22 = Hsapiens[["chr22"]]
```

Now use the `matchPattern` function to find all the locations in which CCGG occurs on chr22.

How many CCGG do we find on chr22?

```{r}
res=matchPattern("CCGG",chr22)
length(res)
```

### Question 2: 

Plot a histogram of the DNA fragment sizes after we cut with CCGG. How would you describe this distribution?

```{r}
size=diff(start(res))
hist(log10(size))
```

### Question 3:

A typical size to filter are DNA regions between 40 and 220 basepairs.

What proportion of the fragments created for chr22 are between 40 and 220 basepairs?

```{r}
mean(size>=40 & size<=220)
```

### Question 4: 

If we sequence all of chromosome 22 we need to sequence 51,304,566 bases. However, if instead we keep only fragments of size between 40 and 220 basepairs, how much sequence would would be left?

```{r}
sum(size[size>=40 & size<=220])
```

### Question 5:

Let's start by the reading in the target information.

```{r}
path="/Users/lucia/OneDrive/Desktop/R/edX_DNA Methylation/colonCancerWGBS-master"
targets <- read.table(file.path(path,"targets.txt"), header = TRUE, sep = "	")
targets
```

Now you will need the `bsseq` package to read in the sequencing data.

We load the methylation calls from our alignments into R. Once the data are loaded into R, we can use this package for further downstream analyses such as finding differentially methylated regions between our paired tumor and normal colon cancer samples. This package assumes the following data has been extracted from the alignments.

1. genomic positions (chromosome and location) for methylation loci

2. `M` (Methylation) values, describing the number of read supporting methylation covering a single loci

3. `Cov` (Coverage) values, describing the number of read supporting methylation covering a single loci

For illustrative purposes, we only consider the methylation loci in chromosome 22 from the `.cov` file.

Here read in the six files:

```{r}
BiocManager::install("bsseq")
library("bsseq")
## turn metadata into DataFrame w/ sample names as rownames
targets <- DataFrame(targets, row.names = as.character(targets$Run))

## specify path to files in same order as targets table
covfiles <- file.path(path, paste0(rownames(targets), ".chr22.cov"))

## read coverage files
colonCancerWGBS <- read.bismark(files = covfiles, rmZeroCov = TRUE,colData = targets)
```

To view the bsseq object and the phenotypic information about each sample:

```{r}
colonCancerWGBS
# phenotypic information
pData(colonCancerWGBS)
# granges object
granges(colonCancerWGBS)
```

Now we can extract the coverage and the number of reads with evidence from methylation:

```{r}
cov=getCoverage(colonCancerWGBS,type = "Cov")
m=getCoverage(colonCancerWGBS,type = "M")
```

What proportion of the reported CpGs have some coverage in all sample?

```{r}
mean(apply(cov,1,min)>0)
#or
index=apply(cov>0,1,all)
mean(index)
```

### Question 6:


Compute the total coverage (across all samples) for each CpG. Plot it against location.

```{r}
totalc=apply(cov,1,sum)
plot(start(granges(colonCancerWGBS)),log10(totalc+1))
abline(log10(300),0,col="red")
abline(log10(200),0,col="blue")
hist(log10(totalc+1))
```

```{r}
# model answer
tot = rowSums(cov)
##there are some very large values
hist(tot)
loc= start(colonCancerWGBS)
##plot by pieces
for(i in 1:11){
  index=1:100000+100000*i ##very ad-hoc
  plot(loc[index],tot[index],cex=.5,ylim=c(0,300))
}
```

### Question 7

Note that we can get coverage and the number of reads including evidence for methylation like this:

```{r}
cov=getCoverage(colonCancerWGBS,type = "Cov")
m=getCoverage(colonCancerWGBS,type = "M")
```

Note that we can now make figures. Here is a selected region:

```{r}
gr = GRanges(seqnames="22",ranges=IRanges(start=43793678,end= 45022550))
index=granges(colonCancerWGBS)%over%gr
library(rafalib)
i=1
index2=which(index & cov[,i]>=5 & cov[,i]<=50)
x=start(colonCancerWGBS)[index2]
y=m[index2,i]/cov[index2,i]
w=sqrt(cov[index2,i])/7
plot(x,y,cex=w)
```